### ISSUE TITLE
The issue of reproduce stage 1

### ISSUE THREAD

**Issue by ziz-797** [wjhou/Radar#3]
Hi, thank you for sharing this great work!

We have followed your dataset setting and trained Stage 1 (i.e., the BLIP-3 backbone model) using 4 × 3090 GPUs. Specifically:

LoRA-finetuned the SigLIP vision encoder and Phi-3.5

Fully finetuned the resampler

All training hyperparameters were exactly the same as in your [run-mimic-cxr.sh](http://run-mimic-cxr.sh/), except batch_size from 2 to 1, due to the OOM issue:

batch_size=1, epochs=3, learning_rate=1e-4

We also followed your provided dataset preprocessing pipeline as discussed previously. However, we encountered a small issue:

After processing the data, during the training process, in your dataset.py, the construct_clinical_context function does not accept float inputs, but some reports have NaN as values for certain keys. This caused a len(val) error. We modified the function slightly to skip any key-value pairs where the value is NaN.

Other than that, we strictly followed your code and preprocessing pipeline.
 However, after training, our evaluation results on the test set are:

BLEU-1: 0.40

BLEU-4: 0.176

ROUGE-L: 0.30

METEOR: 0.35

These results are lower than those reported in your paper, and we are unsure why.

We suspect the discrepancy may come from differences in the training data. Would you be willing to share your processed annotation.json and clinical-context.json files with us? Alternatively, could you please provide more detailed steps or requirements on how to exactly preprocess the data to reach the reported performance?

Any guidance or clarification would be greatly appreciated!

**Comment by wjhou** [comment]
Hi, 

Thanks for reproducing our work.

Here are some suggestions that might be helpful:

1. OOM issue
Four 3090 GPUs with findings truncated to 512 characters should work fine with `batch_size=2`. Note that when `batch_size=1`, you should increase acc_step to ensure the total batch size is 32.

2. Data format
To verify whether the data format aligns with the paper, I suggest printing the input string during debugging. You can add a print statement in [conversation_utils.py](https://github.com/wjhou/Radar/blob/main/src/conversation_utils.py). I suggest checking both the training and test sets using the debug option.

3. Prior Findings
We include prior findings as in MAIRA-2. Please make sure they are included in the context.

Best regards,
Ethan

**Comment by ziz-797** [comment]
OK,your suggestions are really helpful,thank you very much.

**Comment by ziz-797** [comment]
Hi,

Sorry about bothering you for so many times.

After receiving your suggestions,we git a new code from your github and make several changes to the code,which can be concluded below:

1.Delete the KNOWLEDGE_PATH, IMAGE_ANNOTATION_PATH,and GT_OBSERVATION_PATH under the deepspeed in the run_mimic-cxr.sh file.For the reason,we just want to reproduce the result of backbone of your wonderful work,which is stage 1 only.

2.Although our machine is the same 4*3090 as your machine,but we can not run the model unless add the change to optimizer config file zero2.json with the code below:
      "offload_optimizer": {
      "device": "cpu",
      "pin_memory": true
      },
    "offload_param": {
      "device": "cpu",
      "pin_memory": true
      }

3.Replace the Nan float value in the clinical-context.json with "",the clinical-context.json file is produced followed by your work,while if run the original file,it will lead to the error in the dataset.py,for in the function construct_clinical_context the len(val) can not accept the float number.

4.We debug both the testset and trainset and we see the data input is the same as your work.

After we make these works,we get the reproduce result below:

"test_BLEU_1": 0.44027344467969376,

"test_BLEU_4": 0.18155918758029946,

"test_ROUGE_L": 0.30698872749397466,

"test_METEOR": 0.36092585546779965,

As you can see the result is not good.So we want to further ask that:

1:Can you share the environment file or the list of package in your environment? 

For the reason that we can not directly satisfy the environment in your work if using the environment file in the github,we have to adjust many times to let the model run successfully.

2:Can you further provide us with your data?(ANNOTATRION and CLINICAL_CONTEXT)

For our work on dealing with the dataset may be not correct but our work strictly follows the code in your github without any changes.However we can not get the result in your paper.

Thank you very much and looking forward to your reply.

Have a good day!

Best regards

**Comment by wjhou** [comment]
Hi,

The results reported align with the performance of **Backbone-V1**, rather than **Backbone**. I also recommend trying the settings used for **Backbone-V1** (i.e., tuning only the Resampler). 

If your primary goal is to reproduce Stage 1 and you have access to more substantial computational resources (e.g., 8×4090 GPUs), you may consider adjusting the settings to speed up training.

I'm glad to hear that you were able to run the code successfully. The environment should not be a major barrier to reproducing the results.

Best regards,
Ethan

**Comment by ziz-797** [comment]
Hi,

Thank you for your reply.

Just want to confirm,does all the settings in the github are the same  to implement the experiment result of your backbone in the Table 5 of your paper? We have stricty follow the setting in your work without any changes, while we can just get the result showed in the last question.

Thank you very much and have a good day!

**Comment by wjhou** [comment]
Hi, 

I re-ran the code from this repo on a new machine and was able to reproduce the results in Stage 1. Here is a minimum package list to install:
```
conda create -n radar python=3.9.19
conda activate radar
conda install nvidia/label/cuda-12.1.0::cuda-toolkit
conda install pytorch==2.2.1 torchvision==0.17.1 torchaudio==2.2.1 pytorch-cuda=12.1 -c pytorch -c nvidia
pip install transformers==4.41.1
pip install deepspeed==0.14.2
pip install numpy==1.26.3
pip install einops==0.8.0 einops-exts==0.0.4
pip install peft==0.10.0
pip install evaluate==0.4.3
pip install absl-py==2.1.0 rouge-score==0.1.2 nltk==3.8.1
pip install sentencepiece==0.2.0
pip install protobuf==4.25.3
pip install flash-attn==2.5.8 --no-build-isolation
```
This setup has been tested, and you can install it on your machine.

I've fixed some bugs in the repo—please check out the latest code in `src`.

Regarding your experiments, the clinical context data might be affecting your results. Ignoring `nan` values should be acceptable, but I'm uncertain about the correctness of the other information. I recommend directly adopting the clinical context release provided by Libra.

Best regards,
Ethan

**Comment by ziz-797** [comment]
OK,thank you very much and have a good day.It is really helpful!

**Comment by ziz-797** [comment]
Hi,

Thank you very much for your help and it works! The performance increase to:

BLEU-1:  0.4352
BLEU-2:  0.3301
BLEU-3:  0.2663
BLEU-4:  0.2233
ROUGE-L: 0.3733
METEOR:  0.4056

However,we still can not run the model with batch-size 2 in 4*3090 and the performance has not reached to the requirement. It may be caused by other reasons, we have used the corrected clinical-context.Do you have further suggestions?

Thank you very much for your patience and have a good day!

Best regards,

ziz-797

**Comment by wjhou** [comment]
Hi, 

I guess I got it wrong on the 3090 settings, and using batch_size 1 should be fine.

I updated the [conversation_utils.py](https://github.com/wjhou/Radar/blob/main/src/conversation_utils.py). You can integrate it into your code to reproduce the performance.

Best regards,
Ethan

**Comment by ziz-797** [comment]
Hi, 

Thank you for your reply.The performance of us was based on the updated code of your work.Can you provide us with some further suggestions? These suggestions will be very helpful.

Thank you very much and have a good day!

**Comment by wjhou** [comment]
Hi, 

Here are some details I’d like to confirm with you:

1. The conversation format has been updated in `conversation_utils.py`
```
tokenizer.chat_template = "{{ bos_token }}{% for message in messages %}{% if (message['role'] == 'user') %}{{'<|user|>' + '\n' + message['content'] + '<|end|>' + '\n' + '<|assistant|>' + '\n'}}{% elif message['role'] == 'system' %}{{ '<|system|>' + '\n' + message['content'] + '<|end|>' + '\n'}}{% elif (message['role'] == 'assistant') %}{{message['content'] + '<|end|>' + '\n'}}{% endif %}{% endfor %}"
```

2. There are 162,955 training samples and 2,463 test samples in `annotation.json`.

3. The prior findings are included in the context.

Best regards,
Ethan

**Comment by ziz-797** [comment]
Hi,

Good morning,

Thank you for your reply and our dataset is same as you mention and the code is the same.We do not add the prior finding in the clinical context for we see the dataset.py will find the prior finding in the annotation.json and connect to the prompt.We also debug the input to confirm this situation.

Thank you very much and have a good day!

Best regards,
Ziz-797

**Comment by wjhou** [comment]
Hi,

Could you please send me a complete copy of your code? I’ll run it on my server and get back to you afterward.

My email address is `houwenjun060@gmail.com`.

Best regards,
Ethan

**Comment by ziz-797** [comment]
OK,thank you very much and I will email you with the data and code we used,thank you very much and have a good day!

**Comment by wjhou** [comment]
Hi,

I conducted experiments with two settings:
1. Your annotation + your clinical context file
2. Your annotation + my clinical context file

I was able to reproduce the results in both settings. I also reviewed the code you sent to me, and everything looks good. Currently, I don't have any suggestions.

Best regards,
Ethan

**Comment by ziz-797** [comment]
OK,thank you very much,it might be some wrong with our machine and we will check for it.Thank you very much for your patience.

**Comment by ziz-797** [comment]
Hi,

About the missing prior data,I see the data of your wonderful work copy the current image path to the prior image path,which is same as libra.Just want to confirm,do you also introduce the bias in this situation like the Libra to distinguish the prior and current image?

Thank you very much and have a good day!

Best regards,
Ziz-797

**Comment by Tommybear1136** [comment]
@ziz-797 I ran into the same issue as you when trying to reproduce Stage 1. In my case, the problem came from the following settings in zero2.json, which you mentioned earlier:
"offload_optimizer": {
    "device": "cpu",
    "pin_memory": true
},
"offload_param": {
    "device": "cpu",
    "pin_memory": true
}
With these enabled, and using a gradient accumulation step that is four times larger than the author's setting (i.e., 16, since I only have one GPU), I noticed that the gradient norm became four times smaller compared to using a gradient accumulation step of 4. This behavior is also shown in the figure below.

After removing the offloading configurations from zero2.json, the gradient norms became consistent across both settings, and I was able to reproduce the author's results.

Hope this helps!

<img width="482" height="362" alt="Image" src="https://github.com/user-attachments/assets/b1defe2a-a4aa-4b2f-9b0a-996ad881bd11" />
